File: pkg/models/location.go
package models

// LocationType represents location type
type LocationType byte

// Location types
const (
	LOCATION_TYPE_OFFICE     LocationType = 1
	LOCATION_TYPE_WAREHOUSE  LocationType = 2
	LOCATION_TYPE_STORE      LocationType = 3
	LOCATION_TYPE_PRODUCTION LocationType = 4
	LOCATION_TYPE_OTHER      LocationType = 5
)

// Location represents location data stored in database
type Location struct {
	LocationId         int64        `xorm:"PK"`
	Uid                int64        `xorm:"INDEX(IDX_location_uid_deleted_order) NOT NULL"`
	Deleted            bool         `xorm:"INDEX(IDX_location_uid_deleted_order) NOT NULL"`
	CfoId              int64        `xorm:"NOT NULL DEFAULT 0"`
	Name               string       `xorm:"VARCHAR(64) NOT NULL"`
	Address            string       `xorm:"VARCHAR(255) NOT NULL"`
	LocationType       LocationType `xorm:"NOT NULL DEFAULT 1"`
	MonthlyRent        int64        `xorm:"NOT NULL DEFAULT 0"`
	MonthlyElectricity int64        `xorm:"NOT NULL DEFAULT 0"`
	MonthlyInternet    int64        `xorm:"NOT NULL DEFAULT 0"`
	Comment            string       `xorm:"VARCHAR(255) NOT NULL"`
	DisplayOrder       int32        `xorm:"INDEX(IDX_location_uid_deleted_order) NOT NULL"`
	Hidden             bool         `xorm:"NOT NULL"`
	CreatedUnixTime    int64
	UpdatedUnixTime    int64
	DeletedUnixTime    int64
}

// LocationGetRequest represents all parameters of location getting request
type LocationGetRequest struct {
	Id int64 `form:"id,string" binding:"required,min=1"`
}

// LocationCreateRequest represents all parameters of location creation request
type LocationCreateRequest struct {
	Name               string       `json:"name" binding:"required,notBlank,max=64"`
	CfoId              int64        `json:"cfoId,string"`
	Address            string       `json:"address" binding:"max=255"`
	LocationType       LocationType `json:"locationType"`
	MonthlyRent        int64        `json:"monthlyRent"`
	MonthlyElectricity int64        `json:"monthlyElectricity"`
	MonthlyInternet    int64        `json:"monthlyInternet"`
	Comment            string       `json:"comment" binding:"max=255"`
	ClientSessionId    string       `json:"clientSessionId"`
}

// LocationModifyRequest represents all parameters of location modification request
type LocationModifyRequest struct {
	Id                 int64        `json:"id,string" binding:"required,min=1"`
	Name               string       `json:"name" binding:"required,notBlank,max=64"`
	CfoId              int64        `json:"cfoId,string"`
	Address            string       `json:"address" binding:"max=255"`
	LocationType       LocationType `json:"locationType"`
	MonthlyRent        int64        `json:"monthlyRent"`
	MonthlyElectricity int64        `json:"monthlyElectricity"`
	MonthlyInternet    int64        `json:"monthlyInternet"`
	Comment            string       `json:"comment" binding:"max=255"`
	Hidden             bool         `json:"hidden"`
}

// LocationHideRequest represents all parameters of location hiding request
type LocationHideRequest struct {
	Id     int64 `json:"id,string" binding:"required,min=1"`
	Hidden bool  `json:"hidden"`
}

// LocationMoveRequest represents all parameters of location moving request
type LocationMoveRequest struct {
	NewDisplayOrders []*LocationNewDisplayOrderRequest `json:"newDisplayOrders" binding:"required,min=1"`
}

// LocationNewDisplayOrderRequest represents a data pair of id and display order
type LocationNewDisplayOrderRequest struct {
	Id           int64 `json:"id,string" binding:"required,min=1"`
	DisplayOrder int32 `json:"displayOrder"`
}

// LocationDeleteRequest represents all parameters of location deleting request
type LocationDeleteRequest struct {
	Id int64 `json:"id,string" binding:"required,min=1"`
}

// LocationInfoResponse represents a view-object of location
type LocationInfoResponse struct {
	Id                 int64        `json:"id,string"`
	Name               string       `json:"name"`
	CfoId              int64        `json:"cfoId,string"`
	Address            string       `json:"address"`
	LocationType       LocationType `json:"locationType"`
	MonthlyRent        int64        `json:"monthlyRent"`
	MonthlyElectricity int64        `json:"monthlyElectricity"`
	MonthlyInternet    int64        `json:"monthlyInternet"`
	Comment            string       `json:"comment"`
	DisplayOrder       int32        `json:"displayOrder"`
	Hidden             bool         `json:"hidden"`
}

// ToLocationInfoResponse returns a view-object according to database model
func (l *Location) ToLocationInfoResponse() *LocationInfoResponse {
	return &LocationInfoResponse{
		Id:                 l.LocationId,
		Name:               l.Name,
		CfoId:              l.CfoId,
		Address:            l.Address,
		LocationType:       l.LocationType,
		MonthlyRent:        l.MonthlyRent,
		MonthlyElectricity: l.MonthlyElectricity,
		MonthlyInternet:    l.MonthlyInternet,
		Comment:            l.Comment,
		DisplayOrder:       l.DisplayOrder,
		Hidden:             l.Hidden,
	}
}

// LocationInfoResponseSlice represents the slice data structure of LocationInfoResponse
type LocationInfoResponseSlice []*LocationInfoResponse

// Len returns the count of items
func (s LocationInfoResponseSlice) Len() int {
	return len(s)
}

// Swap swaps two items
func (s LocationInfoResponseSlice) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

// Less reports whether the first item is less than the second one
func (s LocationInfoResponseSlice) Less(i, j int) bool {
	return s[i].DisplayOrder < s[j].DisplayOrder
}

File: src/models/location.ts
export class Location implements LocationInfoResponse {
    public id: string;
    public name: string;
    public cfoId: string;
    public address: string;
    public locationType: number;
    public monthlyRent: number;
    public monthlyElectricity: number;
    public monthlyInternet: number;
    public comment: string;
    public displayOrder: number;
    public hidden: boolean;

    private constructor(id: string, name: string, cfoId: string, address: string, locationType: number, monthlyRent: number, monthlyElectricity: number, monthlyInternet: number, comment: string, displayOrder: number, hidden: boolean) {
        this.id = id;
        this.name = name;
        this.cfoId = cfoId;
        this.address = address;
        this.locationType = locationType;
        this.monthlyRent = monthlyRent;
        this.monthlyElectricity = monthlyElectricity;
        this.monthlyInternet = monthlyInternet;
        this.comment = comment;
        this.displayOrder = displayOrder;
        this.hidden = hidden;
    }

    public toCreateRequest(clientSessionId?: string): LocationCreateRequest {
        return {
            name: this.name,
            cfoId: this.cfoId,
            address: this.address,
            locationType: this.locationType,
            monthlyRent: this.monthlyRent,
            monthlyElectricity: this.monthlyElectricity,
            monthlyInternet: this.monthlyInternet,
            comment: this.comment,
            clientSessionId: clientSessionId
        };
    }

    public toModifyRequest(): LocationModifyRequest {
        return {
            id: this.id,
            name: this.name,
            cfoId: this.cfoId,
            address: this.address,
            locationType: this.locationType,
            monthlyRent: this.monthlyRent,
            monthlyElectricity: this.monthlyElectricity,
            monthlyInternet: this.monthlyInternet,
            comment: this.comment,
            hidden: this.hidden
        };
    }

    public clone(): Location {
        return new Location(this.id, this.name, this.cfoId, this.address, this.locationType, this.monthlyRent, this.monthlyElectricity, this.monthlyInternet, this.comment, this.displayOrder, this.hidden);
    }

    public static of(response: LocationInfoResponse): Location {
        return new Location(response.id, response.name, response.cfoId, response.address, response.locationType, response.monthlyRent, response.monthlyElectricity, response.monthlyInternet, response.comment, response.displayOrder, response.hidden);
    }

    public static ofMulti(responses: LocationInfoResponse[]): Location[] {
        const locations: Location[] = [];

        for (const response of responses) {
            locations.push(Location.of(response));
        }

        return locations;
    }

    public static createNew(): Location {
        return new Location('', '', '0', '', 1, 0, 0, 0, '', 0, false);
    }
}

export interface LocationCreateRequest {
    readonly name: string;
    readonly cfoId: string;
    readonly address: string;
    readonly locationType: number;
    readonly monthlyRent: number;
    readonly monthlyElectricity: number;
    readonly monthlyInternet: number;
    readonly comment: string;
    readonly clientSessionId?: string;
}

export interface LocationModifyRequest {
    readonly id: string;
    readonly name: string;
    readonly cfoId: string;
    readonly address: string;
    readonly locationType: number;
    readonly monthlyRent: number;
    readonly monthlyElectricity: number;
    readonly monthlyInternet: number;
    readonly comment: string;
    readonly hidden: boolean;
}

export interface LocationHideRequest {
    readonly id: string;
    readonly hidden: boolean;
}

export interface LocationMoveRequest {
    readonly newDisplayOrders: LocationNewDisplayOrderRequest[];
}

export interface LocationNewDisplayOrderRequest {
    readonly id: string;
    readonly displayOrder: number;
}

export interface LocationDeleteRequest {
    readonly id: string;
}

export interface LocationInfoResponse {
    readonly id: string;
    readonly name: string;
    readonly cfoId: string;
    readonly address: string;
    readonly locationType: number;
    readonly monthlyRent: number;
    readonly monthlyElectricity: number;
    readonly monthlyInternet: number;
    readonly comment: string;
    readonly displayOrder: number;
    readonly hidden: boolean;
}
